### PR TITLE
Allow for reactivity in popover and tooltip contents

### PR DESCRIPTION
**Pull request by jovyntls**
**What is the purpose of this pull request?**

- [ ] Documentation update
- [x] Bug fix
- [x] Feature addition or enhancement
- [ ] Code maintenance
- [ ] Others, please explain:

Relates to #59 

**Overview of changes:**
Allows for reactivity in popover/tooltip contents by:
* Using the vue-bootstrap component instead of directives to implement popovers and tooltips. 
* Using portals for Triggers

**Testing instructions:**
```html
### Popovers

<popover header="header" content="**lorem ipsum** dolor sit amet <panel header='can be clicked'>panel contents</panel>">
hover popover
</popover>

<popover header="Header" content="Lorem ipsum dolor sit amet" trigger="click">
  <button class="btn btn-secondary">Click popover button</button>
</popover>

<br>

This is a <trigger for="pop:trigger_id" trigger="focus">focus trigger</trigger>.
This is a <trigger for="pop:trigger_id">hover trigger</trigger>.
<popover id="pop:trigger_id" content="This popover is triggered by a trigger. <panel header='can be opened'>Lorem ipsum</panel>"></popover>
<br>
This is a <trigger for="pop:trigger_id" trigger="click">click trigger</trigger>.

### Modal

More about <trigger for="modal:loremipsum" trigger="click">trigger</trigger>.
This is the same <trigger for="modal:loremipsum">trigger</trigger> as last one.
<modal header="**Modal header** :rocket:" id="modal:loremipsum">
  Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore
  magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo
  consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
  Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
</modal>

### Tooltips

<tooltip content="**Lorem ipsum** *dolor sit amet* <trigger for='modal:loremipsum'>Modal hover trigger</trigger>." placement="bottom" trigger="click">
  <button class="btn btn-secondary">Tooltip on bottom</button>
</tooltip>
<br />
This is a <trigger for="tt:trigger_id">hover trigger</trigger>.
<tooltip id="tt:trigger_id" content="**Lorem ipsum** *dolor sit amet* <trigger for='modal:loremipsum' trigger='click'>Modal click trigger</trigger>"></tooltip>
<br>
This is a <trigger for="tt:trigger_id">click trigger</trigger> 
```

**Proposed commit message: (wrap lines at 72 characters)**
```
Allow reactivity in popovers and tooltips

Currently, popovers and tooltips are implemented using bootstrap-vue 
directives. 

Such an implementation is not reactive which limits the content that 
can be used in a popover/tooltip. Reactivity is important for future 
enhancements like supporting the src attribute (#59). Furthermore, it 
is incompatible with SSR, resulting in hydration problems that have 
arisen in #1615.

Using Vue components for tooltips and popovers, and Portals for 
triggers, allows us to avoid the issues with reactivity and SSR. 

Let's use components instead of bootstrap-vue directives to implement 
popovers and tooltips. This gives better support of reactivity while 
avoiding SSR issues.
```

---

**Checklist:** :ballot_box_with_check:

- [x] Updated the documentation for feature additions and enhancements
- [ ] Added tests for bug fixes or features
- [x] Linked all related issues
- [x] No unrelated changes

### CODE REVIEW

**Comment by ang-zeyu** [comment]
Thanks for the PR! I think this is in the right direction.

Wonder if we should account for reactivity in popover content inside `<trigger>`s as well in this PR. (maybe even tooltips), since it also changes how content is transfered over to `<popover>`s themselves. (e.g. if you are portalling the content)

**Comment by jovyntls** [comment]
Hi @ang-zeyu , thanks for the suggestion, I think that makes sense especially since portal-ling the content over would remove the need for the data-mb-slot-name attributes in Popover.vue. I'll make the changes to Trigger.vue and Tooltips too!

**Review by jonahtanjz — COMMENTED** [review]
Nice work @jovyntls 👍 Implementation wise looks good.

Found 2 issues while trying it out:

On the documentation site, the popovers and tooltips in the codeAndOutput seems to be bugged. Can refer to the [PR preview](https://deploy-preview-1748--markbind-master.netlify.app/userguide/components/popups#tooltips).

image

All the content of the tooltips and popovers are displayed by the side. Maybe related to this part where the target became `[object Object]`? 
image

The other issue that I noticed is that the content in popovers now "flashes" first before disappearing when the page is loaded initially. This can be quite obvious when there are a few popovers used on the same page and the content contains items that take up quite a bit of space such as panels.

https://user-images.githubusercontent.com/47470981/152503999-53835aca-b26d-46b3-a778-a7029e4c2d20.mp4

I haven't identified what is causing this issue yet.

**Comment by jonahtanjz** [comment]
Also noticed when the popover includes content such as another page, the popover can overflow.
```html
<popover header="header" content="<include src='../index.md' />">
hover popover
</popover>
```

https://user-images.githubusercontent.com/47470981/152506584-5edcf083-2e1a-4f37-b792-ddb7a580064b.mp4

Perhaps we can introduce overflow scroll on the popover itself?

**Comment by jovyntls** [comment]
Thank you for testing @jonahtanjz ! I can't seem to replicate the both issues locally (even with throttling), so currently it seems to be an issue with the deployed site O.o 

Is it okay for me to "spam" commits for the PR preview to test this out?

**Comment by jonahtanjz** [comment]
> I can't seem to replicate the first 2 issues locally (even with throttling), so currently it seems to be an issue with the deployed site O.o

The documentation site on my local setup is fine as well. The "flashes" occurred on my local setup. Will try to use another device to see if this is happening as well.

> Is it okay for me to "spam" commits for the PR preview to test this out? Or should I open a new PR for testing?

Should be fine :)

**Comment by jovyntls** [comment]
I've fixed the issue with the buggy popovers and tooltips in the codeAndOutput — the default slot had to be named `#default`, though I'm not sure why this is so.

Will look into replicating the flashes, and the overflowing behaviour!

**Comment by jonahtanjz** [comment]
> Will look into replicating the flashes, and the overflowing behaviour!

@jovyntls Tried it out on another device and it seems that the flashes are gone. Might have been an isolated issue. Can ignore for now :) The overflowing issue can be replicated.

File: packages/vue-components/src/Popover.vue [hunk 1]
```diff
@@ -0,0 +1,60 @@
+<template>
+  <span
+    :class="trigger === 'click' ? 'trigger-click' : 'trigger'"
+    data-mb-component-type="popover"
+    tabindex="0"
+  >
+    <portal v-if="targetEl.id" :to="targetEl.id">
+      <template #title>
```
Comment (ang-zeyu) [inline packages/vue-components/src/Popover.vue:8]: these are local to the scope of the template being compiled (i.e. `Popover.vue`) (try testing a trigger with a popover with both header and content)

**Review by ang-zeyu — COMMENTED** [review]
There are quite a few hydration errors (in the console) as well locally which may be causing the discepancies (netlify previews are a production build with vue ssr assertions disabled).

File: packages/vue-components/src/Popover.vue (hunk 1 above)
Comment (jovyntls) [reply packages/vue-components/src/Popover.vue]: Thank you for the catch, I've fixed this as well as the hydration errors and checked the console (locally and on the deploy preview) for hydration warnings. Do let me know if I've missed out any other concerns!

File: packages/vue-components/src/Tooltip.vue [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    :class="trigger === 'click' ? 'trigger-click' : 'trigger'"
+    data-mb-component-type="tooltip"
+    tabindex="0"
+  >
+    <portal v-if="targetEl.id" :to="targetEl.id">
+      <slot name="content"></slot>
+    </portal>
+
+    <b-tooltip
+      :target="targetEl"
```
Comment (ang-zeyu) [inline packages/vue-components/src/Tooltip.vue:12]: > [BootstrapVue warn]: tooltip - The provided target is no valid HTML element.

Could we `v-if` this before the target el has initialised as well? (change the default targetEl to `undefined`) (or a `isMounted` flag with `$el`)

Same for popovers

**Review by ang-zeyu — COMMENTED** [review]
Nice work!
I think this fixes #1615 as well 🎉.

Just a few more touchups:
- `data-mb-slot-name` selector in `markbind.css` can probably be removed
- the `'[data-mb-component-type] [data-mb-slot-name]'` selectors in algolia.js too (ssr should take care of this now)

File: docs/userGuide/syntax/popovers.mbdf [hunk 3]
```diff
@@ -79,8 +79,8 @@ This is the same <trigger for="pop:trigger_id">trigger</trigger> as last one.
 Name | Type | Default | Description
 ---- | ---- | ------- | ------
 trigger | `String` |	`hover`	| How the Popover is triggered.<br>Supports: `click`, `focus`, `hover`.
-header{{slot_icon_trigger}} | `String` | `''` | Popover header, supports inline markdown text.
-content{{slot_icon_trigger}} | `String` | `''` | Popover content, supports inline markdown text.
+header{{slot_icon_trigger}} | `String` | `''` | Popover header, supports MarkBind syntax.
+content{{slot_icon_trigger}} | `String` | `''` | Popover content, supports MarkBind syntax.
```
Comment (ang-zeyu) [inline docs/userGuide/syntax/popovers.mbdf:83]: Think we could also allow/fix the typo now that reactivity is possible
```suggestion
content{{slot_info_trigger}} | `String` | `''` | Popover content, supports MarkBind syntax.
```

File: docs/userGuide/syntax/popovers.mbdf [hunk 4]
```diff
@@ -79,8 +79,8 @@ This is the same <trigger for="pop:trigger_id">trigger</trigger> as last one.
 Name | Type | Default | Description
 ---- | ---- | ------- | ------
 trigger | `String` |	`hover`	| How the Popover is triggered.<br>Supports: `click`, `focus`, `hover`.
-header{{slot_icon_trigger}} | `String` | `''` | Popover header, supports inline markdown text.
-content{{slot_icon_trigger}} | `String` | `''` | Popover content, supports inline markdown text.
+header{{slot_icon_trigger}} | `String` | `''` | Popover header, supports MarkBind syntax.
```
Comment (ang-zeyu) [inline docs/userGuide/syntax/popovers.mbdf:82]:
```suggestion
header{{slot_icon_trigger}} | `String` | `''` | Popover header, supports MarkDown text.
```
for consistency with the rest of the ug

File: docs/userGuide/syntax/popovers.mbdf (hunk 3 above)
Comment (jovyntls) [reply docs/userGuide/syntax/popovers.mbdf:83]: Thanks, I didn't realise this!

File: packages/core-web/src/styles/markbind.css [hunk 5]
```diff
@@ -225,6 +220,11 @@ li.footnote-item:target {
     outline: none;
 }
 
+.popover-content {
+    overflow: auto;
+    max-height: 50vh;
+}
```
Comment (jonahtanjz) [inline packages/core-web/src/styles/markbind.css:226]: I believe this css can be moved to `Popover.vue` now that there is a dedicated component for popovers. We can probably shift the css for the triggers as well to `Trigger.vue` but that can be in another PR.

**Review by jonahtanjz — COMMENTED** [review]
Nice work @jovyntls. The overflow issue seems to be fixed 👍 

Just one suggestion, other than that the rest looks good to me.

**Comment by jovyntls** [comment]
Closing to refresh the CI workflow; will reopen

**Review by jonahtanjz — COMMENTED** [review]
Thanks @jovyntls!

Can try to run `npm run lintfix` to resolve the linting issue :)

**Review by jonahtanjz — APPROVED** [review]
LGTM 👍 

Let's wait for @ang-zeyu to approve as well in case he has anything else to add :)

**Comment by ang-zeyu** [comment]
No more comments on the implementation, Lgtm as well! 👍 

> Allow reactivity in popovers and tooltips

Since this change has quite a bit of context, let's try to also summarise in your commit message body the various reasons, and how this pr achieves it:
- Solves part of #1615 
- Reactivity (in so doing also enabling the popover `src` issue)

(a format is available in the pr template's link)
